### PR TITLE
Test using scores

### PR DESCRIPTION
**Pull request by snath-xoc**
Continuation of #13 that I accidentally merged. Preliminary check on using scoring (between predictions and y_test), instead of raw predicted values for conducting statistical test of equivalence. In a sense, this shifts from checking raw predictions to comparing the model's performance itself in distribution. Linked to discussion in https://github.com/snath-xoc/sample-weight-audit-nondet/issues/10 but compared first for transformers.

TO DO:

 

- [x] Try out other scores (so far have checked R2 and mean_squared_error)
- [x]  Update handling of KMeans so that it is not pipelined into Ridge
- [x]  AdaBoostClassifier amongst others are still returning 1 as p-values. I have implemented the shuffling for _resampled which seems to give good enough results but maybe something more is needed, @jeremiedbb any ideas?
- [x]  The test is still a bit finicky, sometimes 15 estimators fail the statistical tests and sometimes 12, this may just be random fluke.
- [x]  check for deterministic using the predictions (not scores)
- [x]  Add flag if deterministic is detected but continue to use kstest
- [x] CategoricalNB still returning error because of unexpected n_features, this only happens under predict_proba
- [x] DBScan predict feature doesn't exist trying fit_predict instead but this is returning unexpected dimensions
- [x] IsotonicRegression predicting NaN values resulting in error when calculating mse, tried to filter out NaN but then all values are gone.

### CODE REVIEW

File: reports/sklearn_estimators_sample_weight_audit_report.ipynb [hunk 1]
```diff
@@ -89,7 +89,8 @@
     "from sklearn.exceptions import ConvergenceWarning\n",
     "import threadpoolctl\n",
     "\n",
-    "\n",
+    "import sys\n",
+    "sys.path.insert(1,'../src/')\n",
     "from sample_weight_audit import check_weighted_repeated_estimator_fit_equivalence\n",
     "from sample_weight_audit.exceptions import UnexpectedDeterministicPredictions\n",
```
Comment (ogrisel) [inline reports/sklearn_estimators_sample_weight_audit_report.ipynb:95]: I think we should stop importing this exception `UnexpectedDeterministicPredictions` now that use a deterministic flag.  We should further simplify the notebook to not pretend we ever run the deterministic test since we do not ever any more.

Comment (ogrisel) [reply reports/sklearn_estimators_sample_weight_audit_report.ipynb:95]: Actually, we can remove the definition of this exception in the library.

File: reports/sklearn_estimators_sample_weight_audit_report.ipynb [hunk 2]
```diff
@@ -921,18 +1152,18 @@
...
-      "✅ 19 passed the deterministic test\n",
-      "❌ 4 failed the deterministic test\n",
-      "✅ 19 passed the statistical test\n",
-      "❌ 17 failed the statistical test\n",
-      "❌ 3 other errors\n",
+      "✅ 0 passed the deterministic test\n",
+      "❌ 0 failed the deterministic test\n",
+      "✅ 28 passed the statistical test\n",
+      "❌ 24 failed the statistical test\n",
+      "❌ 10 other errors\n",
```
Comment (ogrisel) [inline reports/sklearn_estimators_sample_weight_audit_report.ipynb:1166]: We have 7 new "other" errors. There might be a problem.

File: src/sample_weight_audit/estimator_check.py [hunk 3]
```diff
@@ -22,30 +28,28 @@
...
     test_name: str
-    min_p_value: float
-    mean_p_value: float
-    p_values: np.ndarray
-    predictions_weighted: np.ndarray
-    predictions_repeated: np.ndarray
+    p_value: float
+    scores_weighted: np.ndarray
+    scores_repeated: np.ndarray
+    deterministic_flag: bool
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:34]: I think this should be renamed to `deterministic_predictions`.

File: reports/sklearn_estimators_sample_weight_audit_report.ipynb (hunk 2 above)
Comment (snath-xoc) [reply reports/sklearn_estimators_sample_weight_audit_report.ipynb:1164]: Yes some of them are due to the type of inputs e.g. for GammaRegressor:

error with: Some value(s) of y are out of the valid range of the loss 'HalfGammaLoss'.

I suppose one approach is could be switching back to the X, y values used in the deterministic check?

Comment (ogrisel) [reply reports/sklearn_estimators_sample_weight_audit_report.ipynb:1164]: We should rather change the stochastic test to enforce the constraints expressed in the estimator tags as we do inside the deterministic test:

https://github.com/scikit-learn/scikit-learn/blob/1bd24a27498ca8ab4836ebb25ee554f381ae4d8b/sklearn/utils/estimator_checks.py#L1552-L1553

File: reports/sklearn_estimators_sample_weight_audit_report.ipynb [hunk 4]
```diff
@@ -947,11 +1176,11 @@
     "    \"failed the deterministic test\"\n",
```
Comment (ogrisel) [inline reports/sklearn_estimators_sample_weight_audit_report.ipynb:1176]: This code should also be simplified since it's no longer reachable if we no longer run the deterministic test.

File: src/sample_weight_audit/estimator_check.py [hunk 5]
```diff
@@ -80,65 +84,46 @@ def check_weighted_repeated_estimator_fit_equivalence(
...
     # for each dimension.
-    pvalues = []
-    for i in range(stat_test_dim):
-        pvalue = run_1d_test(
-            data_to_test_weighted[i], data_to_test_repeated[i], test_name
-        ).pvalue
-        pvalues.append(pvalue)
-
-    pvalues = np.asarray(pvalues)
+    pvalue = run_1d_test(data_to_test_weighted, data_to_test_repeated, test_name).pvalue
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:118]: Since we never run ndimensional tests anymore, this function could probably be renamed `run_two_sample_test` or `run_statistical_test`.

**Comment by ogrisel** [comment]
> IsotonicRegression predicting NaN values resulting in error when calculating mse, tried to filter out NaN but then all values are gone.

This can probably be fixed by passing `out_of_bands="clip"` in the hparam config for that estimator:

https://scikit-learn.org/dev/modules/generated/sklearn.isotonic.IsotonicRegression.html#:~:text=out_of_bounds,-{%E2%80%98nan%E2%80%99

File: reports/sklearn_estimators_sample_weight_audit_report.ipynb [hunk 6]
```diff
@@ -928,11 +1222,11 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "✅ 19 passed the deterministic test\n",
-      "❌ 4 failed the deterministic test\n",
-      "✅ 19 passed the statistical test\n",
-      "❌ 17 failed the statistical test\n",
-      "❌ 3 other errors\n",
+      "✅ 0 passed the deterministic test\n",
+      "❌ 0 failed the deterministic test\n",
```
Comment (ogrisel) [inline reports/sklearn_estimators_sample_weight_audit_report.ipynb:1226]: This part of the report should also be removed, since we no longer run the deterministic test in this notebook.

**Comment by ogrisel** [comment]
For `CategoricalNB`, we could special case it in the notebook to replace it by `make_pipeline(KBinsDiscretizer(encode="ordinal"), CategoricalNB())` (while keeping `CategoricalNB` for the `est_name`.

**Comment by ogrisel** [comment]
I am fine with merging this PR once the remaining unresolved comments above are addressed.

File: reports/sklearn_estimators_sample_weight_audit_report.ipynb (hunk 1 above)
Comment (snath-xoc) [reply reports/sklearn_estimators_sample_weight_audit_report.ipynb]: Agh sorry it somehow was rebased to have it again, removed now

**Comment by snath-xoc** [comment]
Great most of the fixes are implemented, if it looks O.K. to a second pair of eyes apart from mine we can merge

**Comment by ogrisel** [comment]
Also the notebook needs to be updated to remove the section named "Details on deterministic test errors" because we do not collect any deterministic test result anymore.

The fact that this cell has a non-empty output reveals a problem. Did you press the "restart and run all cells" button before committing the updated notebook? Please do so each time, otherwise the outputs can be inconsistent.

**Comment by ogrisel** [comment]
I pushed a few commits to store the predictions in the results to make it easier to debug low p-values. I also updated the notebook to set the Bonferroni correction based on the number of estimators we test instead of the old statistical dim that is no longer relevant.

**Comment by ogrisel** [comment]
I think I fixed all the remaining bugs related to rounding errors for deterministic models and the handling of clusterers.

**Comment by ogrisel** [comment]
Actually, `GaussianNB` still has problems. Let me fix that.

**Comment by snath-xoc** [comment]
Great thanks for spotting the small data bugs. The changes all make sense to me.

**Comment by ogrisel** [comment]
I don't understand what's going on with `GaussianNB`. If I reduce the number of decimal places (e.g. to 8), the difference between the two prediction sets is still slightly larger than the cut-off level (e.g. slightly larger than 1e-8).

File: src/sample_weight_audit/estimator_check.py [hunk 7]
```diff
@@ -255,50 +282,29 @@ def multifit_over_weighted_and_repeated(
...
+                random_state=seed, **extra_params_weighted
+            )
+            # Use different random seeds for the other group of fits to avoid
+            # introducing an unwanted statistical dependency.
+            est_repeated = clone(est).set_params(
+                random_state=seed + n_stochastic_fits, **extra_params_repeated
+            )
+        else:
+            est_weighted = clone(est).set_params(**extra_params_weighted)
+            est_repeated = clone(est).set_params(**extra_params_repeated)
```
Comment (ogrisel) [inline src/sample_weight_audit/estimator_check.py:306]: Thinking about it, it's a bit of a waste to fit an estimator with no `random_state` parameter 100x times. We should probably optimize this by fitting them only once and using `pvalue = float(np.allclose(score_weighted, score_repeated))` and hard-code the deterministic flag.

**Comment by ogrisel** [comment]
Ok and this also fixed the problem with the `GaussianNB` model (it's a deterministic model).

**Review by ogrisel — APPROVED** [review]
The results in the notebook all look good to me now!
